### ISSUE TITLE
Don't need to save error history for lead-lag

### ISSUE THREAD

**Issue by gunvirranu** [Quest2GM/RobotVision#1]
Hey Sid! I was reading through this, and somethin felt wrong about [saving the history of error terms](https://github.com/Quest2GM/RobotVision/blob/d944c23d927d25065ae4eabcee0e632deadb619c/README.md?plain=1#L151) for the lead-lag controller. That sorta thing wouldn't fly for a real-time system, where [online](https://en.wikipedia.org/wiki/Online_algorithm) algos are generally preferred.

I thought about it for a bit and I think there's a way to do it without that! We can exploit a very nice property of exponentials:
`[A exp(i dt) + B exp((i + 1) dt) + ...] exp(dt)  =  A exp((i + 1) dt) + B exp((i + 2) dt) + ...`
We can keep a running total and then use this property to "shift" the total-so-far back, sample by sample, effectively reducing the weight of all previous samples every time. This allows you to compute the updated convolution every time by essentially relying on previous work. Reduce, reuse, recycle baby! 😉

Here's some sample code I used to verify this. Let me know if I messed anything up and it turns out this doesn't work lol.

```python
import numpy as np
import matplotlib.pyplot as plt

t, dt = np.linspace(0, 10, 1001, retstep=True)
# e(t) can be ANY signal
e = np.sin(2 * t) + 0.7 * np.cos(3 * t)
a, b = 4, 3

# Direct convolution
EXP_WINDOW = (a - b) * np.exp(-b * t)
y1 = np.convolve(e, EXP_WINDOW, mode="full")[:e.size] * dt

# Recursive first-order system
SHIFT_FACTOR = np.exp(-b * dt)
y2 = [0]
for e_i in e:
    y2_new = SHIFT_FACTOR * y2[-1] + (a - b) * dt * e_i
    y2.append(y2_new)
y2 = np.array(y2)[1:]

# They're the same!
assert np.allclose(y1, y2)

# Plot both
plt.style.use("ggplot")
plt.plot(t, e, label=r"$e(t)$")
plt.plot(t, y1, label="Convolution")
plt.plot(t, y2, label="Recursive")
plt.legend()
plt.tight_layout()
plt.show(block=True)
```

**Comment by gunvirranu** [comment]
Ayo @Quest2GM, I got an update for ya. This's been rattling around in my head for a bit, and I *just* realized we literally have the transfer function 🤦‍♀️. Also, it was bothering me that we used a summation to approximate the convolution since there are better ways of numerically estimating definite integrals. After some playing around, it seems the recursive method above is actually similar to a slightly modified implicit Euler method. I just happened to stumble upon a small modification that gives the *exact* same result as the direct summation we used prior. But that's not too important ig.

Anyways, we can actually play around with some better ways of computing this. You can see the code I was messing around with below. It's helpful to have a "truth" value, so I used [`scipy.integrate.quad`](https://docs.scipy.org/doc/scipy/reference/reference/generated/scipy.integrate.quad.html) to compute the convolution from the input function *e(t)* directly (not data samples), which I assumed is the best estimate (`y0`). We can also compare this against [`scipy.signal.lsim`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lsim.html), (`y1`).

The two summation methods from the previous comment are the same, direct convolution (`y2`) and recursive (`y3)`. If we modify the direct convolution to use the [trapezoidal rule](https://en.wikipedia.org/wiki/Trapezoidal_rule) instead (`y4`), we fortunately get a closer approximation. To improve the recursive method though, we gotta resort to a mega-🧠 move and discretize the transfer function *(a - b) / (s + b)* using a [bilinear transform](https://en.wikipedia.org/wiki/Bilinear_transform). I'm not gonna upload my messy math, but I can type it up if needed. Assuming I did it correctly, I get:
`dt (a - b) (e[n-1] + e[n]) + (2 - b dt) y[n-1] = (2 + b dt) y[n]`
Coding this up (`y5`), it turns out it works pretty decently actually. While theoretically this is the same as the previous trapezoidal method, I couldn't get it to match the convolution exactly 🤷‍♀️. In any case, it's still a very good improvement on the previous recursive method, so that's a win in my book.

Just for comparison, I also tried estimating the convolution using [Simpson's rule](https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule) as an improvement over trapezoidal with [`scipy.integrate.simpson`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html), (`y6`). Using the `even="last"` option to keep the error transient towards the left, it works like suuuper well. It's not online though (just like your original summation), so the best applicable solution is still the recursive trapezoidal (`y5`), at least from what I know so far.

Here's the code 🙌. It'll plot the different methods and on the lower plot, show the errors from the actual (`y0`). Should definitely run it to try different *e(t)* functions and see how the errors change. Fun stuff.

```python
import numpy as np
import matplotlib.pyplot as plt
from scipy import integrate, signal

t, dt = np.linspace(0, 10, 1001, retstep=True)
# e(t) can be ANY signal
e_f = lambda x: 0.2 * x + np.sin(2 * x) + 0.7 * np.cos(3 * x)
e = e_f(t)
a, b = 4, 3

# Compute "truth" convolution using e(t) analytical function directly
y0 = []
for t_i in t:
    f_conv = lambda x: e_f(x) * np.exp(-b * (t_i - x)) * (a - b)
    new_y0, _ = integrate.quad(f_conv, t[0], t_i, epsabs=1e-10, epsrel=1e-10)
    y0.append(new_y0)
y0 = np.array(y0)

# SciPy, which should be pretty accurate
tf = signal.TransferFunction([a - b], [1, b])  # TF = (a - b) / (s + b)
_, y1, _ = signal.lsim(tf, e, t)

# Convolution (summation)
EXP_WINDOW = (a - b) * np.exp(-b * t)
y2 = np.convolve(e, EXP_WINDOW, mode="full")[:t.size] * dt

# Recursive (summation)
SHIFT_FACTOR = np.exp(-b * dt)
y3 = [0]
for e_i in e:
    y3_new = SHIFT_FACTOR * y3[-1] + dt * (a - b) * e_i
    y3.append(y3_new)
y3 = np.array(y3)[1:]

# Convolution (trapezoidal)
y4 = np.zeros(t.size)
for i in range(1, t.size + 1):
    y4[i - 1] = np.trapz(e[:i] * EXP_WINDOW[:i][::-1], dx=dt)

# Recursive (trapezoidal)
NEW_SAMPLE_FACTOR = dt * (a - b) / (2 + b * dt)  # Approx. of 1/2 dt (a-b) exp(-1/2 b dt)
SHIFT_FACTOR = (2 - b * dt) / (2 + b * dt)  # Approx. of exp(-b dt)
y5 = [0]
for i in range(1, e.size):
    y5_new = SHIFT_FACTOR * y5[-1] + NEW_SAMPLE_FACTOR * (e[i - 1] + e[i])
    y5.append(y5_new)
y5 = np.array(y5)

# Convolution (Simpson's)
y6 = np.zeros(t.size)
for i in range(1, t.size + 1):
    y6[i - 1] = integrate.simpson(e[:i] * EXP_WINDOW[:i][::-1], dx=dt, even="last")

# Summation convolution and recursive are the same!
assert np.allclose(y2, y3)

# Plot all ze stuff
plt.style.use("bmh")
fig, axs = plt.subplots(2, sharex=True, figsize=(14, 8))
axs[-1].set_xlabel("Time [s]")
axs[0].plot(t, e, label=r"$e(t)$")
axs[0].plot(t, y0, "-+", label="Actual")
axs[0].plot(t, y1, label=r"SciPy $\tt{lsim}$")
axs[0].plot(t, y2, label="Convolution - Sum")
axs[0].plot(t, y3, label="Recursive - Sum")
axs[0].plot(t, y4, label="Convolution - Trapz")
axs[0].plot(t, y5, label="Recursive - Trapz")
axs[0].plot(t, y6, label="Convolution - Simps")
axs[0].legend()
axs[1].set_yscale("log")
axs[1].set_ylabel('|Error from Actual|')
axs[1].plot([], [], [], [])
for y_i in (y1, y2, y3, y4, y5, y6):
    axs[1].plot(t, np.abs(y0 - y_i))
fig.set_tight_layout(True)
plt.show(block=True)
```
![yee](https://user-images.githubusercontent.com/10203710/147975472-7b428f8c-b0c1-4598-b32c-11d718e58b01.png)

**Comment by gunvirranu** [comment]
Also, regarding my first comment on the recursive summation, my explanation of it seems kinda whack reading it back now lol. Just in case it doesn't make sense, I think the best way to understand it is inductively.
If we let `Cᵢ` be the "summed convolution" of `i` points, so

`Cᵢ = e₁ exp(-b Δt i) + ... + eᵢ₋₁ exp(-b Δt 1) + eᵢ exp(-b Δt 0)`,

and then assume we already have computed `Cᵢ`, we can easily get `Cᵢ₊₁` by:

```
Cᵢ₊₁ = Cᵢ exp(-b Δt) + eᵢ₊₁
     = (e₁ exp(-b Δt i) + ... + eᵢ₋₁ exp(-b Δt 1) + eᵢ exp(-b Δt 0)) exp(-b Δt) + eᵢ₊₁
     = e₁ exp(-b Δt (i + 1)) + ... + eᵢ₋₁ exp(-b Δt 2) + eᵢ exp(-b Δt 1) + eᵢ₊₁
```

where the final expression is equivalent to the definition of `Cᵢ₊₁`. This essentially lets use reuse the work from computing previous convolutions because we can easily "update" the previous value using this inductive step. Hope that's a better explanation.

**Comment by Quest2GM** [comment]
Nice work @gunvirranu and thanks for the clever insight! I've updated the theory portion of the README as well as the code to address two things:
1) How we can avoid keeping a running history of error terms.
2) Discretizing our transfer function and then integrating. This is an important step since sampling can never be continuous and it's probably what's done in practice. Also, it's nice because the bilinear transform is the trapezoidal rule in disguise so, as you have pointed out, it carries with it a huge accuracy boost!

**Comment by gunvirranu** [comment]
Sweet update Sid!

> I'll leave coding this up as an exercise to the viewer.

LMAO love it.

A quick thing though. On [this line for the PID control loop](https://github.com/Quest2GM/RobotVision/blob/55ff05464e571348cea023c4f4be9a8e94f2f20f/README.md?plain=1#L72) and [this line for the lead-lag control loop](https://github.com/Quest2GM/RobotVision/blob/55ff05464e571348cea023c4f4be9a8e94f2f20f/README.md?plain=1#L168) of the README, I think it should be `control_output = ...`, not `error = ...`. Funny enough, the `PID.update_gains_PID` method does it correctly cause it doesn't use `self.E` for anything except returning the control output, and it updates the integral with the [correct error](https://github.com/Quest2GM/RobotVision/blob/55ff05464e571348cea023c4f4be9a8e94f2f20f/src/class_main.py#L415) (`self.P`). For `PID.update_gains_LL` however, the [integral update step](https://github.com/Quest2GM/RobotVision/blob/55ff05464e571348cea023c4f4be9a8e94f2f20f/src/class_main.py#L429) should add `self.P` (*not* `self.E`, which is the control output).

**Comment by Quest2GM** [comment]
Damn it, you're absolutely right! Got that taken care of now.

**Comment by gunvirranu** [comment]
@Quest2GM ayo fam, thought you might find this cool! I was curious how linux calculates it's load avg figure (if you run `uptime` for example), so I took a look at the source:

https://github.com/torvalds/linux/blob/master/include/linux/sched/loadavg.h

Turns out, it takes the number of active processes and computes an exponentially weighted moving average, which is basically the same idea to what we did above, except it's not for a lead-lag controller, but only for low-pass filtering. It uses 11-bit fixed point numbers instead, but the `SHIFT_FACTOR` thing from above is pretty much the same. They just hardcoded the values (e.g. `round(exp(-5 / 60) * 2**11) == 1884`, cause it updates every 5 secs), which they use for `EXP_1`. cool stuff
